MRG: add skipmers; switch to reading frame approach for translation, skipmers

bluegenes opened the pull request:
This PR enables skipmers **ONLY in the rust code**.
- enables two skipmer types: m1n3, m2n3
- switches `SeqToHashes` to use reading frame struct, which simplifies/unifies the code across the different methods. The reading frame code handles any modifications needed - i.e. translation or skipping. Then we just kmerize the reading frame as usual. The main difference for translation is that we no longer need to store a buffer of all hashes from the reading frames.

Since this changes the `SeqToHashes` strategy a bit, there's one python test where we now see a different error (modified).


Skipmer References:
- [Skip-mers: increasing entropy and sensitivity to detect conserved genic regions with simple cyclic q-grams](https://www.biorxiv.org/content/10.1101/179960.abstract)
- [Extracting and Evaluating Features from RNA Virus Sequences to Predict Host Species Susceptibility Using Deep Learning](https://dl.acm.org/doi/abs/10.1145/3473258.3473271)

bluegenes committed:
try skipmers

bluegenes committed:
Merge branch 'latest' into try-skipmers

bluegenes committed:
use hash fn encoding instead; init testing

bluegenes committed:
fix test

bluegenes committed:
Merge branch 'latest' into try-skipmers

bluegenes committed:
cont

bluegenes committed:
test

bluegenes committed:
skipmers must be at least ksize 3

mr-eyes commented:
Hi Tessa,
Will you be allowing user-defined n, m,k here? And will you decide to construct the Skipmer after accepting the hash value range or you will construct all Skipmers then hash them and either accept or skip?

bluegenes committed:
Merge branch 'latest' into try-skipmers

bluegenes commented:
> Hi Tessa, Will you be allowing user-defined n, m,k here? 

Hi Mo! At the moment I just got the basics working, and am testing it out over in branchwater. Is there a strong reason for flexible n,m? 

> And will you decide to construct the skipmer after accepting the hash value range or you will construct all Skipmers then hash them and either accept or skip?

By "hash value range" do you mean the FracMinHash selection process (i.e. max hash)? I'm just using the SeqToHashes approach, so my reading of it is that we construct all, then add if it meets the threshold.

Do you have something more efficient/flexible already implemented? And/or what things do you think would be useful here?

bluegenes committed:
clippy fix

mr-eyes commented:
> Hi Mo! At the moment I just got the basics working, and am testing it out over in branchwater. Is there a strong reason for flexible n,m?

Not a strong one. Adding skipmers to sourmash sketching is an excellent addition. So, having the flexibility to change n,m, and k would be good for changing the dispersity/contiguity of the extracted skipmers and, therefore, helping in different applications.

> By "hash value range" do you mean the FracMinHash selection process (i.e. max hash)? I'm just using the SeqToHashes approach, so my reading of it is that we construct all, then add if it meets the threshold.

Gotcha! Just expect that small n,m will have a noticeable slowdown in sketching time. 

> Do you have something more efficient/flexible already implemented? And/or what things do you think would be useful here?

Not really flexible in that context, but you might find the skipmers implementation in kmerDecoder helpful. https://github.com/dib-lab/kmerDecoder/blob/master/src/KD_skipmers.cpp
and this very old example: https://github.com/mr-eyes/OLD_kmerDecoder/blob/d5eb475875ecbe1f3440e1448a56a5ab3b1984fc/python_preview/skipmers.ipynb

bluegenes commented:
> > Hi Mo! At the moment I just got the basics working, and am testing it out over in branchwater. Is there a strong reason for flexible n,m?
> 
> Not a strong one. Adding skipmers to sourmash sketching is an excellent addition. So, having the flexibility to change n,m, and k would be good for changing the dispersity/contiguity of the extracted skipmers and, therefore, helping in different applications.

Got it. I think this shouldn't be too hard if we get a good implementation in, and we could have users specify `m=` and `n=` in the param string. I think I'll probably leave this to the future, but I can try to add `m`,`n` variables in to make future changes easier.

> > By "hash value range" do you mean the FracMinHash selection process (i.e. max hash)? I'm just using the SeqToHashes approach, so my reading of it is that we construct all, then add if it meets the threshold.
> 
> Gotcha! Just expect that small n,m will have a noticeable slowdown in sketching time.

Good point. I haven't done any thinking about optimization yet.

> > Do you have something more efficient/flexible already implemented? And/or what things do you think would be useful here?
> 
> Not really flexible in that context, but you might find the skipmers implementation in kmerDecoder helpful. https://github.com/dib-lab/kmerDecoder/blob/master/src/KD_skipmers.cpp and this very old example: https://github.com/mr-eyes/OLD_kmerDecoder/blob/d5eb475875ecbe1f3440e1448a56a5ab3b1984fc/python_preview/skipmers.ipynb

After reading your implementation, it seems that the main difference is that you take the entire sequence and skipmerize it (remove the skipped bases), then take k-mers/hashes from that sequence as usual. Is that right? Was that a pretty significant speedup compared with just generating skipmers as you go?

mr-eyes commented:
> Got it. I think this shouldn't be too hard if we get a good implementation in, and we could have users specify `m=` and `n=` in the param string. I think I'll probably leave this to the future, but I can try to add `m`,`n` variables in to make future changes easier.

Parameterizing it should be an ideal solution, yes! Thank you!

> After reading your implementation, it seems that the main difference is that you take the entire sequence and skipmerize it (remove the skipped bases), then take k-mers/hashes from that sequence as usual. Is that right? Was that a pretty significant speedup compared with just generating skipmers as you go?

I haven't documented any benchmark here, but I believe I did it that way for performance.

bluegenes commented:
> Parameterizing it should be an ideal solution, yes! Thank you!

Hey Mo! Reading through the 2017 skipmer paper again, they note that triplet (n=3) skipmer patterns performed best, namely m=2,n=3 and m=1,n=3. Do you have a good argument for allowing more patterns than that?

I'm using the hashfunctions (moltype) enum to build sketches and ensure that only compatible sketches are comparable down the road. We don't want incompatible skipmer sketches to be compared. I think unless we currently have evidence to show other combos are useful, I may just enable these two and make them two enums, e.g. `Murmur64Skipm2n3` and `Murmur64Skipm1n3` or similar. I don't think there's any reason we couldn't add more later.

Open to other ideas, though!

bluegenes committed:
Merge branch 'latest' into try-skipmers

mr-eyes commented:
> > Parameterizing it should be an ideal solution, yes! Thank you!
> 
> Hey Mo! Reading through the 2017 skipmer paper again, they note that triplet (n=3) skipmer patterns performed best, namely m=2,n=3 and m=1,n=3. Do you have a good argument for allowing more patterns than that?
> 
> I'm using the hashfunctions (moltype) enum to build sketches and ensure that only compatible sketches are comparable down the road. We don't want incompatible skipmer sketches to be compared. I think unless we currently have evidence to show other combos are useful, I may just enable these two and make them two enums, e.g. `Murmur64Skipm2n3` and `Murmur64Skipm1n3` or similar. I don't think there's any reason we couldn't add more later.
> 
> Open to other ideas, though!

I don't really have a use case in mind for different configurations. So this is good enough for the implementation, and as you said, we could add more later if needed.

bluegenes committed:
WIP: skipmer improvements (#3415)

Make skipmers robust, but keep #3395 functional in the meantime.

This PR:
- enables second skipmer types, so we have m1n3 in addition to m2n3
- switches to a reading frame approach for both translation + skipmers,
which means we first build the reading frame, then kmerize, rather than
building kmers + translating/skipping on the fly
- avoids "extended length" needed for skipping on the fly

Since this changes the `SeqToHashes` strategy a bit, there's one python
test where we now see a different error.

Future thoughts:
- with the new structure, it would be straightforward to add validation
to exclude protein k-mers with invalid amino acids (`X`). I guess I'm
not entirely sure what happens to those atm...

bluegenes committed:
Merge branch 'latest' into try-skipmers

bluegenes commented:
@ctb @mr-eyes @luizirber ready for review.

bluegenes committed:
try fix branchwater incompat

bluegenes committed:
roll back web-sys too

bluegenes committed:
add js-sys, web-sys to dependabot ignore

ctb commented:
On a first pass, looks good to me! I'm not thrilled with the `Murmur64Skipm1n3` abbreviation style and would prefer something longer, but I don't have good suggestions and am not particularly against it, either; so, if you had a longer option in mind that you like, please consider it :)

Now I'm curious how hard it would be to add these to the Python layer 🤔

ctb commented:
@luizirber any concerns, at least a hot-take level?

bluegenes committed:
add more tests

bluegenes committed:
more tests

bluegenes committed:
test protein rf display

bluegenes committed:
Merge branch 'latest' into try-skipmers

bluegenes commented:
ref https://github.com/sourmash-bio/sourmash/issues/659 -- I think might need minor modification for this...

luizirber commented on .github/dependabot.yml:
```diff
@@ -17,6 +17,8 @@ updates:
     - dependency-name: "wasm-bindgen"
     - dependency-name: "once_cell"
     - dependency-name: "chrono"
+    - dependency-name: "js-sys"
+    - dependency-name: "web-sys"
```
Are these off-target? I'm guessing it's because they generate complications downstream on the plugins when updated?

bluegenes replied:
yes, updating these prevented installing with the branchwater plugin :/

luizirber commented on src/core/src/signature.rs:
```diff
@@ -163,27 +164,149 @@ impl SigsTrait for Sketch {
+    pub fn new_skipmer(sequence: &[u8], start: usize, m: usize, n: usize) -> Self {
+        let seq = sequence.to_ascii_uppercase();
+        if start >= n {
+            panic!("Skipmer frame number must be < n ({})", n);
```
Should we `panic!` here and below on `new_translated`, or change the fn signature to return a `Result`?

bluegenes replied:
Would a result be cleaner so we can propagate the error? 

cc @ctb

bluegenes replied:
implemented in #3445, would appreciate a review there

luizirber reviewed (approved):
Awesome! Some minor comments, but thanks for all the test coverage increases =]

bluegenes committed:
Merge branch 'latest' into try-skipmers

bluegenes committed:
seqtohashes errors instead of panic for frame errors (#3445)
